features/events/daemon/src/main/java/org/opennms/netmgt/eventd/EventIpcManagerDefaultImpl.java

```diff
@@ -71,10 +73,52 @@
  * @author <A HREF="http://www.opennms.org">OpenNMS.org </A>
  */
 public class EventIpcManagerDefaultImpl implements EventIpcManager, EventIpcBroadcaster, InitializingBean {
-    
-    
+
     private static final Logger LOG = LoggerFactory.getLogger(EventIpcManagerDefaultImpl.class);
 
+    /**
+     * <p>This queue changes the behavior of {@link #offer(Runnable)} so that it performs a
+     * blocking {@link #put(Runnable)} call instead unless the {@link #offer(Runnable)} is
+     * being invoked by an {@link EventListenerExecutor} thread. In that case, it will attempt
+     * an offer but if that fails because the queue is full, it will instead run the {@link Runnable}
+     * on the current thread. This will prevent deadlocks when {@link EventListener} calls in turn
+     * generate new events and the execution queue is full.</p>
+     * 
+     * <p>This queue is used to throttle the event processing back when the queue is full without 
+     * discarding events.</p>
+     */
+    private static class BlockUnlessInsideEventListenerQueue extends LinkedBlockingQueue<Runnable> {
+
+        private static final long serialVersionUID = 5613562498935257980L;
+
+        public BlockUnlessInsideEventListenerQueue(int capacity) {
+            super(capacity);
+        }
+
+        @Override
+        public boolean offer(Runnable event) {
+            // If we're inside an event listener...
+            if (EVENT_LISTENER_DEPTH.get().intValue() > 0) {
+                // Attempt to offer
+                if (!super.offer(event)) {
+                    // If the queue is full, run on current thread
+                    LOG.warn("Execution queue is full, executing Runnable on current thread");
+                    event.run();
```

Comment:
I assume that you mean to the Eventd prefix so that the log ends up in eventd.log.